Add support for PHP 7.3 and 7.4 - use Twig 3.x, Symfony 5.x and PHPUnit 8.x

mvorisek opened the pull request:
fixes #70, fixes #56

mvorisek commented:
PR complete, tests are passing for PHP 7.1 - 7.4

PierreRambaud commented on .travis.yml:
```diff
@@ -3,8 +3,11 @@ language: php
+  - 7.4
 
 before_script:
+  - if [ "$TRAVIS_PHP_VERSION" == "7.1" ]; then rm composer.lock; fi # remove once PHP 7.1 version is dropped
```
If we remoe the composer.lock, test are not using the same version as we do.
You must remove this condition

mvorisek replied:
> If we remoe the composer.lock, test are not using the same version as we do.
> You must remove this condition

It is not possible to use one `composer.lock` file for PHP 7.1 - 7.4. Is this really an issue, if PHP 7.2 - 7.4 is tested correctly?

mvorisek replied:
@PierreRambaud It is not possible to use one `composer.lock` file for PHP 7.1 - 7.4. Is this really an issue, if PHP 7.2 - 7.4 is tested correctly?

PierreRambaud replied:
I don't understand why it's not possible to have a single possible for both version. If we have different packages versions, this means someone using this tool with 7.4 must remove the composer.lock before trying to use it. :thinking:

mvorisek replied:
Please see https://travis-ci.org/PrestaShop/TranslationToolsBundle/jobs/640871153 . The current lock file makes sense to me, as otherwise the tests will run with very olds libs and the tests will not test if the latests (but still locked) libs are working correctly. By this solution the tests are done with the latest/locked and the old/PHP 7.1 compatible libs as well. Is that fine with you?

PierreRambaud replied:
You need to set up the php version in the composer.lock to be sure it's also compatible with 7.1 when running in 7.4, like we did for https://github.com/PrestaShop/PrestaShop (with php 5.6)

mvorisek replied:
Yes, that is a possible solution, my question is if it is the best solution, as then `twig/twig` package of version 3.0 and `symfony/*` packages of versions 5.0 would not be allowed. What is the real issue with my solution? It tests PHP 7.1 - 7.4 correctly, does not place any limits for newer versions of PHP and it is also fully compatible with PHP 7.1 when this package is installed  thru composer.

PierreRambaud replied:
If the composer.lock needs to be remove to make this project working with a specific version of PHP you have two possibilities:
 * the composer.lock is useless and must be removed
 * target a php version in `config`

mvorisek replied:
Do you fully understand the issue? I can target a php version, but it is ok, that the code will be NOT tested with the latest major version of the libs?

PierreRambaud replied:
The problem with your solution is if someone is using PHP 7.1, he will not be able to install this project because of a bad composer.lock file. The second problem is depending on the php versions you used, dependencies will have different versions, the behavior can also be different because of that.

You should take a look at https://getcomposer.org/doc/06-config.md#platform
If you do that if you try a composer update with php7.4, it will try to be compatible with the platform version AND your php version. This is what we did for the PrestaShop project.

Removing the composer.lock isn't a good idea (IMHO)

mvorisek replied:
> - the composer.lock is useless and must be removed
> - target a php version in config

Both solutions affects only the current project, not if it is used as a dependency. If I will fix the PHP version in composer config, the project will not be tested with the latest Symfony 5 - is it ok?


> The problem with your solution is if someone is using PHP 7.1, he will not be able to install this project because of a bad composer.lock file

So if this project is expected to be used as a dependency, does this reason makes any significant sense?

I agree that lock file is better to be present - wdyt about one extra lock file for PHP 7.1 for the PHP 7.1 CI?

PierreRambaud replied:
I totally forget this project is used as dependency (too many repositories to watch each days :sob:), in this case, the best situation is to remove the composer.lock :)

mvorisek replied:
> in this case, the best situation is to remove the composer.lock :)

done :)

PierreRambaud replied:
Thanks :)

eternoendless commented:
Consider adding `composer.lock` to the `.gitignore` file

mvorisek commented:
> Consider adding composer.lock to the .gitignore file

@eternoendless done & rebased on top of the `master` branch

PierreRambaud commented:
@mvorisek The composer.lock is still in the "Files changed"

mvorisek commented:
> @mvorisek The composer.lock is still in the "Files changed"

Yes, originally I removed it but not excluded from `.gitignore`. Now I have added the lock file back and created it to be installable for PHP 7.1 - 7.4. Does it still needs to be removed?

PierreRambaud commented:
As you said earlier, it's a bundle not an app. You were right when you removed it :+1:

mvorisek commented:
> 
> 
> As you said earlier, it's a bundle not an app. You were right when you removed it 👍

I see, looong time ago... now removed as expected :)

mvorisek commented:
Can this PR be merged?

PierreRambaud commented:
@mvorisek There is a conflict with composer.lock, this PR needs to be rebased

mvorisek commented:
done

Progi1984 commented:
Hello @mvorisek, thanks for your contribution, would you mind to rebase your pull request? Thanks

mvorisek commented:
@Progi1984 Done, please merge immediatelly to prevent further conflicts.

Progi1984 commented:
Thanks @mvorisek
